### PR TITLE
RTT noLoad low pass filter

### PR DESCRIPTION
**Pull request by elandau**
This PR introduces a low pass filter on noLoad RTT to replace the probing function introduced in 0.0.30.  The probing function caused aggressive reduction of the limit resulting in unnecessary load shedding.  The low pass filter lets the algorithm adapt to changes in long-term (10's of minutes) changes noLoad RTT while ensuring that outlier samples don't cause excessive upward drift during prolonged periods of increased latency.

### CODE REVIEW

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/ExpAvgMeasurement.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    private final double ratio;
+    private final double filter;
+    
+    private double value = 0.0;
+    private int count = 0;
+    private final int window;
+    
+    public ExpAvgMeasurement(int window, double filter) {
+        this.ratio = 1.0 / window;
+        this.count = 0;
```
Comment (asibross) [inline concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/ExpAvgMeasurement.java]: You are already setting this value in the field initialization.

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/Measurement.java [hunk 2]
```diff
@@ -11,7 +11,7 @@
      * @param sample
      * @return True if internal state was updated
```
Comment (jhspaybar) [inline concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/Measurement.java]: comment wrong now

Comment (jhspaybar) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/Measurement.java]: Also seems to change the outcome a bit, it's not longer possible to tell if the add actually happened? Does that matter?

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/GradientLimit.java [hunk 3]
```diff
@@ -180,112 +212,84 @@ public static GradientLimit newDefault() {
...
-            return;
         // Normal update to the limit
         } else {
             newLimit = estimatedLimit * gradient + queueSize;
         }
         
-        newLimit = Math.max(queueSize, Math.min(maxLimit, newLimit));
-        newLimit = Math.max(minLimit, estimatedLimit * (1-smoothing) + smoothing*(newLimit));
-        
         if ((int)newLimit != (int)estimatedLimit) {
```
Comment (asibross) [inline concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/GradientLimit.java]: Just wondering whether this int casting may lead to undesirable false outcome (i.e. 20.999.. and 21.000..1 ), perhaps best to compare double with a small epsilon.

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/Measurement.java (hunk 2 above)
Comment (asibross) [inline concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/Measurement.java]: Please fix description of return value.

Comment (elandau) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/Measurement.java]: No, it no longer matters.

**Pull request merged**